python/ray/serve/scheduler.py

```diff
@@ -0,0 +1,124 @@
+import time
+import math
+import ray
+from collections import defaultdict
+from ray.util.scheduling_strategies import NodeAffinitySchedulingStrategy
+
+
+class Scheduler:
+    def __init__(self, deployment_actor_class, spread_min_nodes):
+        # Existing replica placement
+        self.deployment_replicas_topology = {}
+        self.deployment_actor_class = deployment_actor_class
+        # This can be changed to other spread constraints we want to support
+        self.spread_min_nodes = spread_min_nodes
+
+    def add_replica(self):
+        while True:
+            cluster_view = ray._private.state.state.cluster_view()
+            target_node_id = self._schedule(
+                cluster_view,
+                self.deployment_replicas_topology,
+                self._deployment_actor_resources(),
+                self.spread_min_nodes,
+            )
+            if target_node_id is not None:
+                replica = self.deployment_actor_class.options(
+                    scheduling_strategy=NodeAffinitySchedulingStrategy(
+                        node_id=target_node_id, soft=False, _fail_on_unavailable=True
+                    )
+                ).remote()
+                try:
+                    ray.get(replica.__ray_ready__.remote())
+                    self.deployment_replicas_topology[replica] = target_node_id
+                    break
+                except ray.exceptions.ActorUnschedulableError:
+                    # Retry
+                    pass
```

Comment:
this part needs to go through the regular serve control loop, can't be done synchronously like this

Reply:
but we can just re-try to schedule it when we get this error